Added HomeWork 4.

wowwooollie opened the pull request:
Task 2 and 3 have tests written with pytest, while tasks 3 and 4 were tested with doctest.

wowwooollie committed:
Added HomeWork 4.
Task 2 and 3 have tests written with pytest, while tasks 3 and 4 were tested with doctest.

Arisha96 commented on hw4/task03.py:
```diff
@@ -0,0 +1,8 @@
+
+def my_precious_logger(text: str):
+    if text.split()[0][:5].upper() == 'ERROR':
+        print(text, file=sys.stderr)
```
easier and more readable would be using function .startswith can be applied to any str

Arisha96 commented on hw4/task05.py:
```diff
@@ -0,0 +1,22 @@
+            yield 'buzz'
+        else:
+            yield f"{i}"
+
```
as I remember you saw more optimize way for this task

Arisha96 commented on hw4/tests/test_task02.py:
```diff
@@ -0,0 +1,30 @@
+    with open(path, 'w') as f:
+        f.write('AND THE BEST ONE\nIS DEFINITELY "THE GREATEST SHOWMAN"')
+    with pytest.raises(ValueError):
+        read_magic_number(path)
```
one of the point of tests for this task was creating and deleting temporary files, as I see you have only creating
moreover the ideal solution would be to do it via setup() and teardown() fixtures

Arisha96 commented:
also don't see task with filter, where you musted to test existing function

wowwooollie committed:
Changed HomeWork 4.
Task 2 and 3 have tests written with pytest, while tasks 3 and 4 were tested with doctest.

Changed solutions for tasks 3,4 and 5 according to your comments.
Changed test for task 2, but instead of using setup&teardown fixtures I've written a static function,
because I can't figure out how to use parametrized fixtures.
I need them parametrized because the fixture must rewrite the file with different text for each test.

And I haven't tested the first task because 'the faulty case' is the example case. The script doesn't seem to work. So I probably think that there's no need in testing not working scripts.
